File: job_message_writer/backend/app/schemas/resume_tailor.py
# File: backend/app/schemas/resume_tailor.py
from enum import Enum
from pydantic import BaseModel
from typing import Dict, List, Optional, Any


class ResumeSection(str, Enum):
    SKILLS = "SKILLS"
    PROJECTS = "PROJECTS"
    EXPERIENCE = "EXPERIENCE"


class SectionContent(BaseModel):
    original_content: str
    optimized_content: str


class ResumeTailorRequest(BaseModel):
    job_description: str
    resume_id: Optional[int] = None  # If not provided, use active resume
    sections_to_optimize: List[ResumeSection]


class ResumeTailorResponse(BaseModel):
    original_ats_score: float
    optimized_ats_score: float
    optimized_sections: Dict[ResumeSection, SectionContent]


class ATSScoreRequest(BaseModel):
    job_description: str
    resume_content: str


class ATSScoreResponse(BaseModel):
    score: float
    breakdown: Dict[str, float]
    suggestions: List[str]


class SectionOptimizationRequest(BaseModel):
    section_type: ResumeSection
    section_content: str
    job_description: str


class SectionOptimizationResponse(BaseModel):
    optimized_content: str


class PDFOptimizeRequest(BaseModel):
    job_description: str
    resume_id: Optional[int] = None


class TextChange(BaseModel):
    section: str
    type: str
    original: str
    optimized: str


class PDFOptimizeResponse(BaseModel):
    download_id: str
    sections_found: List[str]
    sections_optimized: List[str]
    original_ats_score: float
    optimized_ats_score: float
    changes: List[TextChange] = []


class PDFSectionMapResponse(BaseModel):
    total_spans: int
    total_lines: int
    sections: List[Dict[str, Any]]

File: job_message_writer/frontend/src/lib/api/resumeTailor.ts
import api from "@/lib/api";

export interface SectionInfo {
  name: string;
  content_text: string;
  num_content_lines: number;
  char_count: number;
}

export interface SectionMapResponse {
  total_spans: number;
  total_lines: number;
  sections: SectionInfo[];
}

export interface TextChange {
  section: string;
  type: string;
  original: string;
  optimized: string;
}

export interface PDFOptimizeResponse {
  download_id: string;
  sections_found: string[];
  sections_optimized: string[];
  original_ats_score: number;
  optimized_ats_score: number;
  changes: TextChange[];
}

export interface ATSScoreResponse {
  score: number;
  breakdown: Record<string, number>;
  suggestions: string[];
}

export async function getSectionMap(
  jobDescription: string,
  resumeId?: number
): Promise<SectionMapResponse> {
  const { data } = await api.post("/resume-tailor/section-map", {
    job_description: jobDescription,
    resume_id: resumeId,
  });
  return data;
}

export async function optimizePDF(
  jobDescription: string,
  resumeId?: number
): Promise<PDFOptimizeResponse> {
  const { data } = await api.post("/resume-tailor/optimize-pdf", {
    job_description: jobDescription,
    resume_id: resumeId,
  });
  return data;
}

export function getDownloadUrl(downloadId: string): string {
  const base = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000/api";
  return `${base}/resume-tailor/download/${downloadId}`;
}

export async function getATSScore(
  resumeContent: string,
  jobDescription: string
): Promise<ATSScoreResponse> {
  const { data } = await api.post("/resume-tailor/score", {
    resume_content: resumeContent,
    job_description: jobDescription,
  });
  return data;
}
